pandas/tests/test_categorical.py

```diff
@@ -4549,6 +4549,13 @@ def test_concat_categorical(self):
                             'h': [None] * 6 + cat_values})
         tm.assert_frame_equal(res, exp)
 
+    def test_rank_categorical(self):
+        exp = pd.Series([1., 2., 3., 4., 5., 6.], name='A')
+        dframe = pd.DataFrame(['first', 'second', 'third', 'fourth', 'fifth', 'sixth'], columns=['A'])
+        dframe['A'] = dframe['A'].astype('category', ).cat.set_categories(
+                ['first', 'second', 'third', 'fourth', 'fifth', 'sixth'], ordered=True)
+        res = dframe['A'].rank()
+        tm.assert_series_equal(res, exp)
```

Comment:
@ikilledthecat you will need to check for orderedness in the `_get_data_algo`. Because for non-ordered, we don't want to use the order of the categories, so ranking based on the codes will not be correct.

It may be faster to actually recompute the codes (something like `values.set_categories(values.categories.sort_values()).codes`) and still use the integer codes instead of taking the object path for this case.